src/_pytest/warnings.py

```diff
@@ -81,6 +81,25 @@ def warning_record_to_str(warning_message: warnings.WarningMessage) -> str:
         warning_message.lineno,
         warning_message.line,
     )
+    if warning_message.source is not None:
+        try:
+            import tracemalloc
```

Comment:
Are there platforms which don't have ``tracemalloc`` available?

Reply:
It is likely that PyPy doesn't support it at all due to differing memory management.